### ISSUE TITLE
ai_xgb eviction: add bench xgb-equiv-evict for on-device vs trainer numerical equivalence

### ISSUE THREAD

**Issue by johnjezl** [SLM-OS/SLM-Operating-System#932]
## Motivation

The scheduler cascade has `bench xgb-equiv` (issue #904, PR #921) which replays a verification corpus through the on-device walker and compares against Python's `TripleClassifier.predict`. That verb caught a real divergence — #920, 23/1000 match — that we would otherwise have shipped.

The XGBoost **eviction** policy uses the same `runtime/src/ml/xgb_tree.rs` engine but has no equivalent equivalence check. "XGBoost eviction matches Python" is currently an assumption, not a measurement. Given that XGBoost eviction is a strong candidate for THE default eviction policy in SLM-OS based on its performance numbers, that assumption needs verification.

## Scope

Add `bench xgb-equiv-evict <vectors-path> <expected-path>` (or similar name) mirroring the scheduler verb:

- Routes FAT (`0:/...`) and VFS (`/mnt/files/...`) paths via existing `runtime_blob_read_fat_buf` / VFS helpers.
- Replays N input vectors through `rust_eviction_blob_predict` (or whatever the activation-side entrypoint becomes).
- Compares per-vector output against an `expected_*.bin` corpus produced by the trainer.
- Reports match-rate, first-mismatch index, ns/predict.
- Returns 0 on PASS, 1 on FAIL.

Corpus generation on the trainer side: extend `scripts/export_models.py` (sibling repo) to emit `test_vectors_xgb_evict.bin` + `expected_evict.bin` alongside the model blob.

## Acceptance

- `bench xgb-equiv-evict 0:/slmstore/test_vectors_xgb_evict.bin 0:/slmstore/expected_evict.bin` reports `N/N match PASS` on pi-5-2 (and ideally also on the synthetic-cascade test in `runtime/src/lib.rs::rust_run_tests`).
- Documentation entry in `docs/benchmarks.md` matching the existing scheduler section.
- A small embedded corpus (≤16 vectors) added to `rust_run_tests` so future regressions fail in `make test` rather than on hardware.

## Relationship to #920

If #920's root cause turns out to be in the shared `xgb_tree` engine (candidate 3: f32/f64 threshold drift), this verb is **required to verify the eviction fix**. If #920's root cause is cascade-specific (candidates 1, 2, 4 — leading suspicion), this verb is **forward-looking insurance** confirming the eviction path is clean and catching future regressions.

Either way, the verb is worth landing. Do this *after* #920 resolves so the eviction equivalence test isn't bootstrapped on top of an unresolved scheduler-side bug.

## Out of scope

- Retraining or evaluating eviction model quality (covered by sibling-repo work).
- Changing the eviction blob format.
- Promoting XGBoost to default eviction policy (separate decision, depends on this verb passing).

## Related

- #920 — scheduler-side divergence (blocks the "is the shared engine clean?" question)
- #904 / PR #921 — scheduler `bench xgb-equiv` (template for this work)
- #58 — XGBoost scheduler umbrella

**Comment by johnjezl** [comment]
## Scope split — producer side done, consumer side handed off to #448

The producer side of this issue lives in a separate repo than I assumed when filing — the eviction XGBoost trainer + exporter is in `slm-os-page-eviction`, not `slm-os-scheduler-ai`. The runtime SEMB-loader for eviction blobs existed but had no producer; nothing in either trainer repo emitted SEMB blobs for eviction.

That's now landed as **slm-os-page-eviction PR #3** (`xgb-evict-smb-exporter`):
- New `src/export/xgb_to_smb.py` writes the SEMB+XGB1 blob using SLM-OS's current eviction wire format.
- Base-score fold prepends a synthetic single-leaf tree carrying `logit(base_score)` so on-device tree-sum matches Python's `sigmoid(base_margin + sum_of_trees)` — same mechanism as the scheduler-side fix landed for #920 (slm-os-scheduler-ai PR #4).
- `generate_evict_verification_corpus` emits `test_vectors_xgb_evict.bin` and `expected_evict.bin` from the trained model.
- `scripts/export_to_slmos.py` gains `--smb-output-dir` / `--smb-corpus-size` flags so a single export run produces both the baked `.rs` and the blob+corpus.
- 5 new pytest cases (20/20 total in `test_export.py`).
- Smoke-tested end-to-end against the trained model: 238 KB `evict.smb` plus 100-vector corpus alongside the existing 1.3 MB baked file.

## Consumer side — left to bundle with #448

I'm explicitly **not** doing the SLM-OS-side consumer work in this issue's branch:

- New FFI `rust_eviction_xgb_predict(features: *const f32, len: usize, out_score: *mut f32) -> i32`
- New shell verb `bench xgb-equiv-evict <vectors-path> <expected-path>`
- New shell verbs `eviction blob load/activate xgboost <path>` if missing
- `rust_run_tests` kernel-side regression test against an embedded mini-corpus
- Docs entry in `docs/benchmarks.md` matching the scheduler-side section

All five touch the same `runtime/src/mm/eviction/blob.rs` / `rust_eviction_blob_*` / `XGBoostPolicy` surface that #448 is reworking. Writing them against today's interface would force a mechanical rewrite once #448 lands. Whoever closes #448 should pick this scope up — the new interface shape will be visible to them in a way it isn't to me.

## How the handoff should look from the #448 side

1. Consume the slm-os-page-eviction PR #3 corpus + `.smb` blob as test fixtures (both are independent of envelope changes — they're producer-side bytes the new parser must accept either way).
2. Implement the FFI + shell verb + regression against the new envelope/interface from #448.
3. Hardware-verify on pi-5-2: stage the new `evict.smb`, activate, run `bench xgb-equiv-evict 0:/slmstore/test_vectors_xgb_evict.bin 0:/slmstore/expected_evict.bin`, confirm N/N match PASS (analogous to the #920 1000/1000 result for the scheduler).
4. Close this issue.

## Tolerance note

The verb should compare per-vector sigmoid output within a small tolerance (~1e-4 absolute) rather than bit-equality, since the on-device tree walker stores f32 leaves and applies sigmoid in f32, while Python's `booster.predict()` may use slightly higher-precision intermediates. The scheduler verb compares integer labels (no tolerance needed); eviction is different.

## Related

- slm-os-page-eviction PR #3 — producer side (this comment's subject)
- #448 — envelope generalization (consumer side bundle target)
- #952 — eviction model graduation pathway docs + safety nets
- #920 — scheduler-side equivalence verb (template the consumer side mirrors)

**Comment by johnjezl** [comment]
## Reopening — closing criterion not met

Closed at 2026-05-17T05:56:29Z, one second after PR #955 merged at 05:56:28Z. No hardware verification comment was posted. The issue's own closing criterion (set in the prior comment) was:

> Hardware verification (pi-5-2)
> Required before closing #932:
> 1. Stage the new `evict.smb` to `/slmstore/`.
> 2. Activate the XGBoost eviction policy.
> 3. Run `bench xgb-equiv-evict 0:/slmstore/test_vectors_xgb_evict.bin 0:/slmstore/expected_evict.bin`.
> 4. Confirm **N/N match PASS** (analogous to the 1000/1000 result the scheduler verb produced for #920).

Step 4 has not happened. The `rust_run_tests` mini-corpus added to #955 covers FFI plumbing in QEMU, which is good — but it uses a Rust-built 1-tree synthetic model, not the producer-side `evict.smb` from slm-os-page-eviction PR #3. Python-vs-on-device numerical equivalence on the real trained model has not been demonstrated yet.

## What's also still pending

While reopening — a couple of follow-ups that should be folded into the hardware-run PR or split out:

- **Stale `TODO(#448 handoff)` markers** at `runtime/src/lib.rs:2170`, `:2258`, `kernel/include/slm_ffi.h:463`, and `kernel/src/shell_sys.c:1412`. All say "once #448 lands, this body likely collapses to a single registry call." #448 has landed (closed 2026-05-17T05:56:30Z). Either do the envelope-onto refactor or remove the TODOs as obsolete.
- **`docs/benchmarks.md:528`** still has three result rows marked `*(deferred — pending #448)*`, with prose at L522 saying "the moment the new envelope lands." Wording is stale regardless of the hardware result.

## Closing criterion (unchanged)

The hardware run on pi-5-2 against the producer-side fixtures. Once that passes and the docs row is populated with a real match-rate / latency / first-mismatch result, this can close.

## Related

- #448 (closed — envelope rework landed)
- PR #955 (merged — consumer-side FFI + verb + QEMU mini-corpus regression)
- slm-os-page-eviction PR #3 (merged — producer-side `evict.smb` + corpus)
- #953 (xgb-as-default decision gate — downstream of this closing)
- #952 (graduation pathway docs — downstream of this closing)

**Comment by johnjezl** [comment]
## Verification PASS — 1000-vector run

PR #959 (`4ae7382c`) closed this issue via "Closes #932". Posting the final hardware result here for the timeline:

**pi-5-2 (BCM2712 Cortex-A76 @ 2.4 GHz), 2026-05-17:**
- Match rate: **1000 / 1000** (tolerance 1e-4)
- Per-decision latency: **1,407,823 ns (~1.41 ms)**
- First mismatch: **none**

This 10× larger corpus run confirms the 100-vector hardware result that landed inline with #955 (commit dbd2ec37) — the gap when #932 originally closed was that the verification had happened but wasn't commented here. Both runs now stand as independent confirmations.

## What landed across the bundle

1. Producer side — `slm-os-page-eviction` PR #3 (merged): SEMB+XGB1 exporter, base-score fold, verification-corpus generator. Produces `evict.smb` + corpus from the trained XGBoost model.
2. Consumer side — #955 (merged): `rust_eviction_xgb_predict` / `rust_eviction_xgb_predict_compare` FFIs, `bench xgb-equiv-evict` shell verb, `rust_run_tests` mini-corpus regression for QEMU coverage, initial docs section.
3. Closing run — #959 (merged): 1000-vector hardware verification, docs row updated, four stale `TODO(#448 handoff)` markers removed (since #448 closed without those refactors landing).

## Reproduce

```
python3 ~/projects/slm-os-page-eviction/scripts/export_to_slmos.py \
  --xgb-model ~/projects/slm-os-page-sim/data/models/xgb_model.json \
  --smb-output-dir <out> --smb-corpus-size 1000

# stage evict.smb + corpus to /slmstore/ on the Pi 5 card, then in SLM-OS shell:
eviction model load xgboost 0:/slmstore/evict.smb
eviction model activate xgboost
bench xgb-equiv-evict 0:/slmstore/test_vectors_xgb_evict.bin \
                      0:/slmstore/expected_evict.bin
```

## Downstream now unblocked

- #953 (xgb-as-default decision gate)
- #952 (graduation pathway docs + safety nets)
- #957 (`EVICTION_MODELS=ON` build break) — separate P3 in the page-sim generator, not gating

The parse-per-call latency overhead inside `rust_eviction_xgb_predict_compare` is now noted in `docs/benchmarks.md` as latency context rather than a TODO. If it becomes load-bearing for some future workflow, file fresh.
